File: src/domain/messages.js
import {
  normalizeMessageMode,
  parseMessageModeSectionHeader,
} from './messageModes.js';

export function parseMessageEntries(config) {
  const entries = [];
  let currentMode = null;

  for (const rawLine of String(config ?? '').split(/\r?\n/)) {
    const line = rawLine.trim();
    if (!line) continue;
    const sectionMode = parseMessageModeSectionHeader(line);
    if (sectionMode) {
      currentMode = sectionMode;
      continue;
    }
    if (line.startsWith('#') || !currentMode) continue;
    const commaIndex = line.indexOf(',');
    if (commaIndex <= 0) continue;
    const keyAndLabel = line.slice(0, commaIndex).trim();
    const target = line.slice(commaIndex + 1).trim();
    const parts = keyAndLabel.split(/\s+/, 2);
    const key = String(parts[0] ?? '')
      .trim()
      .toUpperCase();
    if (!key.startsWith('F')) continue;
    entries.push({
      mode: currentMode,
      key,
      label: String(parts[1] ?? '').trim(),
      target,
    });
  }

  return entries;
}

export function actionFromTemplate(template) {
  const match = String(template ?? '')
    .trim()
    .match(/^\{\s*action\s*:\s*([^}]+?)\s*\}$/i);
  return match ? match[1].trim() : null;
}

export function messageActionForConfig(config, mode, key) {
  const entry = messageEntryForConfig(config, mode, key);
  return entry ? actionFromTemplate(entry.target) : null;
}

export function messageEntryForConfig(config, mode, key) {
  const normalizedMode = normalizeMessageMode(mode);
  const normalizedKey = String(key ?? '')
    .trim()
    .toUpperCase();
  if (!normalizedKey) return null;
  return (
    parseMessageEntries(config).find(
      (candidate) =>
        candidate.mode === normalizedMode && candidate.key === normalizedKey,
    ) ?? null
  );
}

function messageFieldText(fields, key) {
  const value = fields?.[key];
  if (value === undefined || value === null) return '';
  return String(value).trim();
}

function cutNumberText(value) {
  return String(value ?? '')
    .trim()
    .toUpperCase()
    .replaceAll('9', 'N');
}

export function renderMessageTemplate(template, fields = {}) {
  return String(template ?? '')
    .replace(/\{([A-Z][A-Z0-9_]*)\}/g, (_match, key) => {
      if (key === 'RST_SENT' || key === 'SENTRSTCUT') {
        return cutNumberText(fields.RST_SENT);
      }
      return messageFieldText(fields, key);
    })
    .trim();
}

export function renderMessageForConfig(config, mode, key, fields = {}) {
  const entry = messageEntryForConfig(config, mode, key);
  if (!entry || actionFromTemplate(entry.target)) return null;
  return renderMessageTemplate(entry.target, fields);
}

export function messageSendBatches(messages, separateMessages = false) {
  if (messages.length === 0) return [];
  if (separateMessages) {
    return messages.map(({ key, text }) => ({ keys: [key], text }));
  }
  return [
    {
      keys: messages.map(({ key }) => key),
      text: messages.map(({ text }) => text).join(' '),
    },
  ];
}

export function completionTrackedTextRequest(requestId, text) {
  return {
    request_id: requestId,
    text,
    wait_for_completion: true,
  };
}

File: radio-io/src/messages.rs
use crate::message_mode::{RUN_MESSAGE_MODE, parse_message_mode_section_header};
use std::collections::HashSet;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ParsedMessageEntry {
    pub mode: String,
    pub key: String,
    pub label: String,
    pub target: String,
}

pub fn parse_message_entries(config: &str) -> Vec<ParsedMessageEntry> {
    let mut entries = Vec::new();
    let mut current_mode = None::<&str>;

    for raw_line in config.lines() {
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }

        if let Some(mode) = parse_message_mode_section_header(line) {
            current_mode = Some(mode);
            continue;
        }
        if line.starts_with('#') {
            continue;
        }

        let Some(entry) = parse_message_line(line, current_mode) else {
            continue;
        };
        entries.push(entry);
    }

    entries
}

pub fn validate_message_config(config: &str, target_label: &str) -> Result<(), String> {
    let mut current_mode = None::<&str>;
    let mut run_keys = HashSet::new();
    let mut search_and_pounce_keys = HashSet::new();

    for (index, raw_line) in config.lines().enumerate() {
        let line_number = index + 1;
        let line = raw_line.trim();
        if line.is_empty() {
            continue;
        }

        if let Some(mode) = parse_message_mode_section_header(line) {
            current_mode = Some(mode);
            continue;
        }
        if line.starts_with('#') {
            continue;
        }

        let mode = current_mode
            .ok_or_else(|| format!("line {line_number}: message is outside a mode section"))?;
        let entry = parse_message_line(line, Some(mode))
            .ok_or_else(|| format!("line {line_number}: expected 'F# Label,{target_label}'"))?;
        if !is_valid_function_key(&entry.key) {
            return Err(format!(
                "line {line_number}: message key must be F1 through F12"
            ));
        }

        let keys = if mode == RUN_MESSAGE_MODE {
            &mut run_keys
        } else {
            &mut search_and_pounce_keys
        };
        if !keys.insert(entry.key.clone()) {
            return Err(format!(
                "line {line_number}: duplicate {} message in {mode} section",
                entry.key
            ));
        }
    }

    if run_keys.is_empty() {
        return Err("messages must include at least one Run message".to_string());
    }
    if search_and_pounce_keys.is_empty() {
        return Err("messages must include at least one S&P message".to_string());
    }

    Ok(())
}

pub fn action_from_template(template: &str) -> Option<String> {
    let inner = template.trim().strip_prefix('{')?.strip_suffix('}')?.trim();
    let (name, value) = inner.split_once(':')?;
    if name.trim().eq_ignore_ascii_case("action") {
        Some(value.trim().to_string())
    } else {
        None
    }
}

fn parse_message_line(line: &str, current_mode: Option<&str>) -> Option<ParsedMessageEntry> {
    let (key_and_label, target) = line.split_once(',')?;
    let mut parts = key_and_label.splitn(2, char::is_whitespace);
    let key = parts.next()?.trim();
    let label = parts.next().unwrap_or("").trim();
    let normalized_key = key.to_uppercase();
    if !normalized_key.starts_with('F') {
        return None;
    }

    Some(ParsedMessageEntry {
        mode: current_mode.unwrap_or_default().to_string(),
        key: normalized_key,
        label: label.to_string(),
        target: target.trim().to_string(),
    })
}

pub fn is_valid_function_key(key: &str) -> bool {
    matches!(
        key.trim().to_uppercase().as_str(),
        "F1" | "F2" | "F3" | "F4" | "F5" | "F6" | "F7" | "F8" | "F9" | "F10" | "F11" | "F12"
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    const TEST_MESSAGES: &str = r#"
# RUN Messages
F1 Cq,CQ TEST
F12 Clear,{Action:Clear}
# S&P Messages
F1 Qrl?,QRL?
"#;

    #[test]
    fn parses_entries_and_action_tokens() {
        let entries = parse_message_entries(TEST_MESSAGES);
        assert_eq!(entries.len(), 3);
        assert_eq!(entries[0].mode, "run");
        assert_eq!(entries[0].key, "F1");
        assert_eq!(entries[0].label, "Cq");
        assert_eq!(
            action_from_template(&entries[1].target),
            Some("Clear".to_string())
        );
    }

    #[test]
    fn validates_message_config_sections_and_duplicates() {
        assert!(validate_message_config(TEST_MESSAGES, "Message").is_ok());
        assert!(validate_message_config("F1 Bad,CQ", "Message").is_err());
        assert!(
            validate_message_config(
                "# RUN Messages\nF1 A,CQ\nF1 B,CQ\n# S&P Messages\nF1 C,CQ",
                "Message"
            )
            .is_err()
        );
    }
}
